### PR TITLE
Prodmon PanelPlot 

### PR DESCRIPTION
**Pull request by shawnlewis**
Timeseries improvements for PanelPlot:
- Timeseries binning ops, so you can setup PanelPlots that automatically rebin when you zoom in
- Fixes to zooming and mutations when x_domain is timestamps. Synchronized zooming on time axis now works
- Don't zoom on an axis if its a sum/count aggregation (see comments in PanelPlot.tsx)
- Better client side mutation support
- Some improvements to SelectEditor (not used in this PR)
- Two notebooks: MonitorPanelPlot and MonitorPanelPlotGeneral. We can reorganize these later.

### CODE REVIEW

File: weave-js/src/components/Panel2/PanelPlot/PanelPlot.tsx [hunk 1]
```diff
@@ -2277,16 +2284,19 @@ const PanelPlot2Inner: React.FC<PanelPlotProps> = props => {
...
               type: xAxisType,
             };
-            newSpec.encoding.x2 = {
-              field: fixedXKey + '.stop',
-              type: xAxisType,
-            };
+            // newSpec.encoding.x2 = {
+            //   field: fixedXKey + '.stop',
+            //   type: xAxisType,
+            // };
```
Comment (dannygoldstein) [inline weave-js/src/components/Panel2/PanelPlot/PanelPlot.tsx:2299]: ?

File: weave-js/src/components/Panel2/PanelPlot/PanelPlot.tsx [hunk 2]
```diff
@@ -2936,18 +2950,39 @@ const PanelPlot2Inner: React.FC<PanelPlotProps> = props => {
                 const currentSetting =
                   currConcreteConfig.signals.domain[dimName];
 
+                let newDomain: Node;
+                if (
+                  PlotState.getAxisType(config.series[0], dimName) ===
```
Comment (dannygoldstein) [inline weave-js/src/components/Panel2/PanelPlot/PanelPlot.tsx:2955]: need to use `currConfig` here, or the reference will go stale

File: weave-js/src/components/Panel2/PanelPlot/PanelPlot.tsx [hunk 3]
```diff
@@ -2936,18 +2950,39 @@ const PanelPlot2Inner: React.FC<PanelPlotProps> = props => {
...
+                  //   a: callOpVeryUnsafe('op-date_parse', {
+                  //     dt_s: constString(new Date(axisSignal[0]).toISOString()),
+                  //   }) as any,
+                  //   b: callOpVeryUnsafe('op-date_parse', {
+                  //     dt_s: constString(new Date(axisSignal[1]).toISOString()),
+                  //   }) as any,
+                  // } as any);
+                } else {
+                  newDomain = constNode(toWeaveType(axisSignal), axisSignal);
+                }
```
Comment (dannygoldstein) [inline weave-js/src/components/Panel2/PanelPlot/PanelPlot.tsx:2972]: why is this needed?

File: examples/app/MonitorPanelPlot.ipynb [hunk 4]
```diff
@@ -0,0 +1,398 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "5e0d6245",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import numpy as np\n",
+    "import pandas as pd\n",
+    "import random\n",
+    "from faker import Faker\n",
```
Comment (dannygoldstein) [inline examples/app/MonitorPanelPlot.ipynb:45]: ensure this gets installed in `pip install '.[examples]'`

File: examples/app/MonitorPanelPlot.ipynb [hunk 5]
```diff
@@ -0,0 +1,398 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "039f298a",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import pandas as pd\n",
+    "import weave\n",
+    "weave.use_frontend_devmode()\n",
```
Comment (dannygoldstein) [inline examples/app/MonitorPanelPlot.ipynb:22]: we should strip this line out of example notebooks

File: examples/app/MonitorPanelPlot.ipynb [hunk 6]
```diff
@@ -0,0 +1,398 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "a672c782",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import tqdm\n",
```
Comment (dannygoldstein) [inline examples/app/MonitorPanelPlot.ipynb:149]: unenumerated dependency. maybe instead of adding to requirements we could do a `pip install` line at the beginning of the notebook to install the notebook's dependencies

**Review by dannygoldstein — COMMENTED** [review]
- we should definitely retain the cross-series line tooltip for line plots in panelplot. right now it seems to disappear on zoom. for example, try zooming one of the second row plots in `MonitorPanelPlotGeneral.ipynb` and mousing over the resulting canvas. there is no longer a line tooltip. it was busted before but is working now. without it you just get a tooltip for a single point in a single line, and its not clear what series its coming from. it was causing crashes before but isn't anymore. can we add that back?

Screen Shot 2023-06-28 at 1 14 18 PM

- colors change on zoom (try zooming the upper left hand corner plot in `MonitorPanelPlotGeneral.ipynb`, which is a bit jarring - we should fix the colors based on the data values when there isn't a color scale specified, so that they stay constant across zooms

**Comment by dannygoldstein** [comment]
More comments here

https://www.loom.com/share/969b61eff4ef4f0faf367b4791e7c3c0?sid=e34b2752-ecc7-42bc-852d-26748bf37c7e

**Review by tssweeney — COMMENTED** [review]
Retroactive Approval
